### ISSUE TITLE
Change the package architecture

### ISSUE THREAD

**Issue by XescuGC** [cycloidio/goat#14]
In issue #12 we updated the documentation to `go get github.com/cycloidio/goat/auth0` to install the package.

## Now
Basically the problem was that the root of the project does not have any imported code, instead all the code is in `/auth0`, which is not `go friendly`.

## Proposal
We should move all the code form `/auth0` to the root of the package to make the package more consistent and clear with go starndars.
But this proposal involves a bracking change, the import path would no longer be `github.com/cycloidio/goat/auth0` so all the projects with this import path will have an import mismatch (if they are updated), because `github.com/cycloidio/goat/auth0` will no loger exist. So this proposal brakes **COMPATIBILITY**.

**Comment by ifraixedes** [comment]
From my point of view and to avoid breaking the current users of the package, I would create a new repository and I would freeze this one writing a _deprecation_ notice in the README.

Another possibility, if you don't like to have a new repo with a new import path, is to copy all the sources of _auth0_ to the root of the project. I wrote "copy", not "move" because we can keep the current version in such directory to avoid to break the current users of the package. If we do this, we'll have to write a proper explanation in the README and we may want to create an issue mentioning this and designate a date (in a year???) that we're going to remove the subdirectory so all the users should migrate their implementations.

**Comment by XescuGC** [comment]
I thought about proposing to move all the code to the root but this kind of repetition is `harmful`, but when you mention `deprecation` I remembered [semver](http://semver.org/) and this is what they say about `deprecations` http://semver.org/#how-should-i-handle-deprecating-functionality:

> How should I handle deprecating functionality?
>
>Deprecating existing functionality is a normal part of software development and is often required to make forward progress. When you deprecate part of your public API, you should do two things: (1) update your documentation to let users know about the change, (2) issue a new minor release with the deprecation in place. Before you completely remove the functionality in a new major release there should be at least one minor release that contains the deprecation so that users can smoothly transition to the new API.

So I have another proposal, (should I open another issue?).

### Use semver

There are a lot of tools from Vendoring in Go because Google uses monorepo so it's not an issue for them, but recently, an OSS project has the "bless" (not quite yet but) that it may be incorpored on the `go tools`. This project is [`dep`](https://github.com/golang/dep) and you can find the documentatio of how they resolve versioning [here](https://github.com/golang/dep/blob/master/FAQ.md#how-does-dep-decide-what-version-of-a-dependency-to-use).

So basically the other proposal is to **add taggs** to the repository for versioning. This way we can `deprecate` the functionality according to the versioning, so in a new `major` we can remove the `/auth0`.

**Comment by ifraixedes** [comment]
Because `dep` isn't yet the standard tool, and [`go get` cannot handle git tags other than `go1`, defaulting to main stream if it doesn't exist](https://golang.org/cmd/go/#hdr-Download_and_install_packages_and_dependencies), and although the Go community has mostly adopted to vendor the dependencies I prefer not to make assumptions, besides that this package may be use for another library (no executable) and [libraries shouldn't vendor the dependencies](https://github.com/golang/go/issues/15162)(at least for now if we consider to be stuck to the official Go tooling) so for those reason I'm suggesting to  give a grace period for keeping the current sources under the _auth0_.

Of course, repetition is **harmful**, but we are not talking about of having to maintain the sources which live in _auth0_ directory, just leave them there for awhile, adding a deprecation message in the README and we may consider to do one of the following changes in the current sources:

* Add some prints to the console when using the old sources in order to make the warning more clear.
* [Use custom imports paths](https://docs.google.com/document/d/1jVFkZTcYbNLaTxXD9OcGfn7vYv5hWtPx9--lTx1gPMs/edit) for the current sources in _auth0_ directory.
 Although using custom imports paths will produce an error as removing the sources, the error message could help to find out easily what happened, when using `go get`
  Example of an error message if we delete the sources:
  ```
   main.go:6:2: cannot find package "github.com/cycloidio/goat/auth0" in any of:
        /root/apps/go/src/github.com/cycloidio/goat/auth0 (from $GOROOT)
        /root/workspace/beego/src/github.com/cycloidio/goat/auth0 (from $GOPATH)
  ```

  Example of an error message if we use the comments for _custom imports path_
  ```
   main.go:6:2: code in directory /root/workspace/beego/src/github.com/cycloidio/goat/auth0 expects import "github.com/cycloidio/goat"
  ```

**We are not repeating because we don't know how to have an abstraction of the logic for being used from different places**, we are duplicating because we don't want to break any current user of the package.

**Comment by ifraixedes** [comment]
I pushed a [branch](https://github.com/cycloidio/goat/tree/if-14-poc) with a proof of concept of copying the sources and setting the _custom imports paths_ in the current sources.

**Comment by XescuGC** [comment]
Looks perfect! But it looks starange not to have a file in the root of the package that not have the name of the package, but the olny one could be the `call.go` but then it should initialize some object called `goat` to make sense hehe. In the other hand I would rename the `call.go` to `auth0.go` because it make sense as it holds the `Auth0` struct.

**Comment by ifraixedes** [comment]
Such change looks good to me.

Could you handle it? I'm busy. If you can, then  just take my branch or create a new one from it and make such change and send a PR

**Comment by XescuGC** [comment]
Yes sure, I do it in your `if-14-poc`?

**Comment by ifraixedes** [comment]
as you wish

**Comment by XescuGC** [comment]
Pushed to the branch, we can create the PR (you or me 😄 )

**Comment by XescuGC** [comment]
PR #16
